### DISCUSSION TITLE
shutdown RKE2 cluster

### DISCUSSION THREAD

**Discussion by mcguppy** [rancher/rke2#2761] [Q&A]
We have the use case to be able to shutdown an RKE2 cluster (all masters and all workers) in case of an power outage during the time the emergency power from batterie is still available (about 10 minutes).

What is the correct procedure to shutdown a whole cluster?
I tried to do a `systemctl stop rke2-server rke2-agent` first on all workers  and after about 30 sec to do a `shutdown -h now` on all workers. Then I did the same on all master nodes. 
All the master nodes shut down correctly, but the workers get stucked in shutting down.
After about 30 minutes, I did a hard power down on all workers.
When switching on all again, the cluster comes back online and all apps as well.
But I need a solution to really shutdown all workers the correct way, so I don't have to do it the hard way.

Is there any correct procedure for this? I have searched the web for quite a long time and the only doc I found was [https://www.suse.com/support/kb/doc/?id=000020031](url)
This is not possible as I do not have docker as container engine but containerd.

Thank you very much for any help on this.

**Comment by briandowns** [comment]
That seems like a reasonable procedure. When you say that the workers are hung, is it the RKE2 process that's hanging? If so, what is in the logs at the time of the shutdown/hang?

**Reply by mcguppy** [reply]
After the `shutdown -h now` command, I just see on the console messages from pods that still access the PVs (we are using Longhorn as storrage provider on the workers).
I haven't analyse these messages that much to be honest, as I thought, this process of shutting down isn't correct at all. But if you think, this is correct, I have to analyse this in more detail the next days when I am back in office.
Thank you very much.

**Reply by Martin-Weiss** [reply]
One challenge I see with this procedure is that the containers stay up running even when stopping rke2-server/rke2-agent. This has the advantage, that you can upgrade with less / no downtime for the workload.. but during reboot the containers might hang depending on their storage dependencies etc.. There is the „rke2-killall“ script you can execute before shutdown to stop everything..

I believe it would also be good to stop the workload before stopping rke2-server/agent (scale replicasets to 0..).

**Reply by mcguppy** [reply]
Hi Martin-Weiss
Thank you very much for this tip. I will try this `rke2-killall` script next week when I am back in the office lab. I hope with this it is working. 
Scaling down replicas to 0 could be quite difficult to perform in the short time we will have  emergency power. But if `rke2-killall` is not doing the job, we will try this as well.
Thanks again for all your help.

**Reply by mcguppy** [reply]
I was able to test it this week several times and with `rke2-killall.sh` followed by `shutdown -h now` about 30 sec later, first on all worker nodes and then on the masters, all went down quite quickly. For a 7 Node cluster it takes me about 2 - 3 minutes, what should be possible for the emergency power from batteries.
Bring the whole cluster back up worked also without problems (at least on my tests this week). It takes some time until all pods are up again, especially the once with PVs attached. Also in Longhorn, there are some replicas, which had to be rebuild, but giving enought time to all (20 - 30 minutes), all was up and running agiain.
Thank you very much for all your help.

**Reply by kkmaslowski** [reply]
@Martin-Weiss cordoning and draining nodes is not needed before shutdown? 
Should I always start all rke2 servers in parallel or one by one? I have setup with 3 servers and 2 agent nodes.
And one more question - when  I have 3 servers and try to shutdown all machines one by one does it not end with lost etcd quorum? Last machine will notice that 2 from 3 servers are down.

**Reply by Martin-Weiss** [reply]
@kkmaslowski - IMO the best approach for „shutdown a cluster“ is to stop the workload before shutting down the cluster. In case of a single server shutdown a cordon / drain is required.. In case no workload is running - the cordon / drain would not be required. ETCD should recover properly in case quorum is lost and nodes come back. So sequential or parallel shutdown does not matter. In case you have hyper converged storage in the cluster (longhorn or ceph or similar) this needs additional attention during shutdown / startup.

**Reply by kkmaslowski** [reply]
@Martin-Weiss thanks for reply. I'm using out-of-tree vsphere cpi. Maybe this is my main problem with shutting cluster. One of servers is missing in etcd after reboot of all machines. I only see error on it ''the member has been permanently removed from the cluster'.

**Reply by Martin-Weiss** [reply]
@kkmaslowski  - in one of my clusters I am also running vsphere cpi/csi and I do not have a problem with etcd... so assuming this might be something to look more into details on why you see this error.. etcd. is basically a local instance and I do not see a good reason why a member should be removed with shutdown / startup "automatically".. Do you have static IP / names or do they change during shutdown/startup i.e. with dynamic DHCP?

**Reply by kkmaslowski** [reply]
@Martin-Weiss we are using static IPs for servers. I'll try to reproduce error. I restarted servers few times and still first server (master node- installed as first) cannot join etcd cluster. It also dissapeared from kubectl list nodes.

**Reply by Martin-Weiss** [reply]
@kkmaslowski  - how did you deploy the cluster? Do you have a load balancer VIP for server= / for port 9345 and 6443?

**Reply by kkmaslowski** [reply]
@Martin-Weiss setup is almost the same as described here: https://computingforgeeks.com/deploy-kubernetes-on-rocky-using-rke2/
3 servers, 2 agents, load balancer. Now I see one difference - on second and third server in file /etc/rancher/rke2/config.yaml I have "server: https://server1.example.com:9345". I should set here address of balancer, not first server.The same for agents.

Please clarify - after installing all servers should I add 'server:' and 'token' also on first server? Both options should be added on all servers?

Server that failed in my setup is server1. It had no 'server:' and 'token:' options, it was installed as first.

**Reply by Martin-Weiss** [reply]
@kkmaslowski - yes - server: should go to the load balancer for :9345 (which goes to the three masters) and after installation you should add the server: also on the first one.

**Reply by kkmaslowski** [reply]
@Martin-Weiss thanks for answers. We will reconfigure it as you described.

**Reply by bovy89** [reply]
can you provide more details about that here? https://github.com/rancher/rke2/discussions/3701
